Check python major version, add shell code py3 compat.

aneeshusa opened the pull request:
Start to prepare salt-ssh for py3 compatibility (not complete).

Update the python code in the shell script to not choke on py3 when checking installed versions.

Current blocker on py3 compat is PyYAML: the origin host runs py2, so it will always send py2 modules versions to targets. However, the py2 PyYAML module uses old-style relative imports, which do not work in py3. I haven't gotten a chance to get past this yet, so there may be more blockers.

The workaround in this PR is to require a python major version match on the original host and the target; the last update to the py2 PyYAML module was in late Nov 2014, so I'll look into fixing this upstream. A full solution would probably involve checking the remote python version and sending the correct modules (or having all modules be fully cross py2/py3 compatible).

aneeshusa committed:
Check python major version, add shell code py3 compat.

Start to prepare salt-ssh for py3 compatibility:
 - Require python major version match:
    - PyYAML (and possibly other modules) is not cross version compatible
      'import yaml' chokes due to old-style relative imports in PyYAML
      Last py2 PyYAML update was Dec 2014, will look into fixing upstream
    - Update error message
 - Prevent python version check from choking on py3:
    - Update exec, print to be function calls
    - Update decoding for new unicode strings
    - Remove extraneous encoding declaration in shim module
      - unneeded by PEP8
      - breaks utf-8 decode on py2

rallytime commented:
Go Go Jenkins!

jfindlay commented:
@aneeshusa, some of these lint errors should be due to old lint problems on the develop branch.  Can you rebase your `salt-ssh-py3` branch on the latest develop?  I've just tested this and there are no conflicts, thanks.

aneeshusa commented:
FYI, the pylint failure is because I removed the utf-8 encoding string in the shim file. I updated the decoding to be py2/py3 compatible, and py2 chokes on the encoding string. I wasn't able to find a way to get Python 2 to accept it with the encoding string (aside from stripping that line), so I just took it out of the file.

s0undt3ch commented:
Py2 chockes on that line?! That's really weird...

aneeshusa commented:
It's been a while since I did this, so I'll take another look. Seemed really weird to me as well!

aneeshusa commented:
After checking into this again, it looks like python2 will throw a "SyntaxError: encoding declaration in Unicode string", the same error as I when I previously investigated. The decoding process is actually fine, but when calling exec python2 doesn't like the encoding statement. There have been [previous problems](https://bugs.python.org/issue22221) with this, which apparently don't exist in python3. That particular bug was fixed in python2, and after some testing, in general encoding strings don't seem to trigger the error. We seem to trigger some pathological edge case :(.

In case anyone wants to look into this further, I've extracted the [compiled version of the shim from this branch](https://gist.github.com/aneeshusa/c1799b69119b44662ca2). I did split it into multiple lines so it's clear that the exec causes the error. Stripping the encoding string is the only solution I've found to this.

thatch45 commented:
Ok this all looks good, thanks for the explanation on the encoding line.We just need to add a lint ignore in there

aneeshusa committed:
Add lint ignore for missing encoding in shim file.
